system/Publisher/ContentReplacer.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+/**
+ * This file is part of CodeIgniter 4 framework.
+ *
+ * (c) CodeIgniter Foundation <admin@codeigniter.com>
+ *
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+namespace CodeIgniter\Publisher;
+
+use RuntimeException;
+use stdClass;
+
+class ContentReplacer
+{
+    /**
+     * Replace content
+     *
+     * @param array $replaces [search => replace]
+     */
+    public function replace(string $content, array $replaces): string
+    {
+        return strtr($content, $replaces);
+    }
+
+    /**
+     * Add text
+     *
+     * @param string $text    Text to add.
+     * @param string $pattern Regexp search pattern.
+     * @param string $replace Regexp replacement including text to add.
+     *
+     * @return stdClass {updated: bool, content: string}
```

Comment:
I'm not opposed to the monad return, but it doesn't seem to be necessary since `content` is irrelevant when `updated === false`. This ends up being functionally the same as `?string`, where `null` means "not updated".

Reply:
`$result->updated` always shows it is updated or not.
`$result->content` always has the content.

Yes, we can use `?string`, but we need to check if it is null, and if not we update the content.

Personally, I think return values with changing meanings, such as `?string`, are a bit difficult to understand.

Reply:
I like the concept of monads (see recent Slack article and thread) but we don't use them anywhere else in the entire framework. In this particular case it increases clarity but doesn't alter efficiency - you are either checking `$result->updated === false` or `$result === null`. 

If this is a development direction you think the framework should take I would rather lean into it and create a framework Monad class that we could use everywhere. If you want to use this here I would prefer to keep these returns internal, either with a docblock or changing the scope on these methods.

Reply:
I should add more personally: I'm pretty new to the concept of monads. I do think they can add a lot of clarity to returns but they definitely come with some functional overhead and they obscure things for new developers. `?string` is very clear that about what the return choices are here, even if there is not a clear mapping of type to "intent"; using `stdObject` or `Monad` requires "unpacking" the object to understand what type is inside it and (for someone who might be seeing it for the first time) how to determine success/failure. 

For these reasons I tend to think monads are more successful on large projects with either private code or a consistent team, where their strengths can be used without any of the weaknesses. For smaller, quicker, or public projects I prefer the simpler return types and either using `null` for "failure"/"no data" or using a straight type and throwing an exception that is meant to be caught. Both of these approaches accomplish the same thing functionally and are easier to understand if you are seeing them the first time in your IDE or docblock.

Reply:
I did not intend to recommend using Monad in the framework.
It was just a Result object, but it seems to use stdClass is not good practice.
Because the structure of stdClass object is not clear.

I've changed the all return types to `?string`, because there is less code to be changed.